Solutions/Corvus.Monitoring.Instrumentation.Abstractions/Corvus/Monitoring/Instrumentation/OperationInstanceExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿// <copyright file="OperationInstanceExtensions.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Monitoring.Instrumentation
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Extension methods for <see cref="IOperationInstance"/>s.
+    /// </summary>
+    public static class OperationInstanceExtensions
+    {
+        /// <summary>
+        /// Adds a property to the additional data associated with this request.
+        /// </summary>
+        /// <param name="operation">The operation to add detail to.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        public static void AddOperationProperty(this IOperationInstance operation, string name, object value)
+        {
+            if (value is null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            operation.AddOperationProperty(name, value.ToString()!);
+        }
+
+        /// <summary>
+        /// Adds one or more entries to the additional data associated with this request.
+        /// </summary>
+        /// <param name="operation">The operation to add detail to.</param>
+        /// <param name="detail">The detail to add.</param>
+        public static void AddOperationDetail(this IOperationInstance operation, AdditionalInstrumentationDetail detail)
```

Comment:
I had assumed this was going to remain in `IOperationInstance`, and that you'd add new members instead of converting the existing member into an extension method.

The problem with moving this out of `IOperationInstance` into an extension method is that this breaks binary compatibility even for code that is only a consumer of `IOperationInstance` (which, as far as we know, is basically all existing code using this library).

Adding new members to an existing interface doesn't break existing consumers of that interface. It's only a breaking change for types that attempt to implement the interface. But if you remove existing members, then that breaks binary compat.

Turning an interface method into an extension method is source-level compatible, but by breaking binary compatibility we've created a DLL hell type situation where there didn't need to be one. Consider some application that didn't even know it was using `Corvus.Monitoring` because it's doing so via its dependencies. If one of its dependencies uses `Corvus.Monitoring` v1, and another uses v2, the modern .NET SDK generally just unifies this to a single version (unless the components are being used in different assembly load contexts), and typically v2 will win. If `IOperationInstance` had retained `AddOperationDetail` as an interface member, that would be OK: any code that was expecting v1 will expect `IOperationInstance` to have just an `AddOperationDetail` method, and it will work happily against v2. But that won't work if you've turned it into an extension method. In that case, you'll get JIT compile time missing method exceptions when the JIT attempts to compile code that was built against v1.

You can no longer just upgrade to `Corvus.Monitoring` v2 at the app level. Your app will need all of the components it uses to agree on which `Corvus.Monitoring` version to use, which will mean those components will need to release a new version compiled against `Corvus.Monitoring` v2.

This is all allowed under semver of course—there's no guarantee of compatibility across major version numbers. But semver doesn't _require_ you to create a situation like this. Since we have the option to make this change breaking only for `IOperationInstance` implementers, and to preserve binary-level compatibility for `IOperationInstance` consumers, I think we should.

I think this would actually be a reasonable use of default interface implementation by the way. That would avoid forcing all implementors to write the old dictionary-flavoured version. (The only reason not to would be if we were supporting .NET Standard, but since we were already .NET >= Core 3.1 only, and v2 will be .NET >= 6 only, then we have this option.) In principle we could even make this a binary-compatible change by adding a default interface implementation for the new method, but I think that might be a mistep, because I think we want to make your newly added method the primary one. (I briefly wondered whether we could define default implementations for both the old and new methods, each implemented in terms of the other, so implementors could implement either the old or the new style and it would just work, and it would be binary compatible with v1. However, I think it would also be very confusing to anyone who didn't know the background, and would also have the unfortunate effect that the default implementation would just recurse infinitely.)

Reply:
Yes that's fine. I'd originally intended to use a default interface implementation but was thwarted by the fact we were targeting netstandard2.0. After a chat with @mwadams I updated to net6.0 but didn't revisit the approach. Have moved the old method to being a default interface implementation.